cmd/snap-confine/seccomp-support.c

```diff
@@ -17,869 +17,182 @@
-	if (real_uid == 0 || effective_uid == 0 || saved_uid == 0)
-		if (seccomp_attr_set(ctx, SCMP_FLTATR_CTL_NNP, 0) != 0)
-			die("Cannot disable nnp");
-
-	// Note that secure_gettenv will always return NULL when suid, so
-	// SNAPPY_LAUNCHER_SECCOMP_PROFILE_DIR can't be (ab)used in that case.
-	if (secure_getenv("SNAPPY_LAUNCHER_SECCOMP_PROFILE_DIR") != NULL)
-		filter_profile_dir =
-		    secure_getenv("SNAPPY_LAUNCHER_SECCOMP_PROFILE_DIR");
+	debug("loading bpf program for security tag %s", filter_profile);
 
 	char profile_path[512];	// arbitrary path name limit
-	sc_must_snprintf(profile_path, sizeof(profile_path), "%s/%s",
+	sc_must_snprintf(profile_path, sizeof(profile_path), "%s/%s.bpf",
 			 filter_profile_dir, filter_profile);
 
-	f = fopen(profile_path, "r");
-	if (f == NULL) {
-		fprintf(stderr, "Can not open %s (%s)\n", profile_path,
-			strerror(errno));
-		die("aborting");
-	}
-	// Note, preprocess_filter() die()s on error
-	preprocess_filter(f, &pre);
-
-	if (pre.unrestricted) {
-		seccomp_release(ctx);
-		ctx = NULL;
-		goto out;
-	}
-	// FIXME: right now complain mode is the equivalent to unrestricted.
-	// We'll want to change this once we seccomp logging is in order.
-	if (pre.complain) {
-		seccomp_release(ctx);
-		ctx = NULL;
-		goto out;
-	}
-
-	char buf[SC_MAX_LINE_LENGTH];
-	while (fgets(buf, sizeof(buf), f) != NULL) {
-		lineno++;
-
-		// skip policy-irrelevant lines
-		if (validate_and_trim_line(buf, sizeof(buf), lineno) == 0)
-			continue;
-
-		char *buf_copy = strdup(buf);
-		if (buf_copy == NULL)
-			die("Out of memory");
-
-		int pr_rc = parse_line(buf_copy, &sargs);
-		free(buf_copy);
-		if (pr_rc != PARSE_OK) {
-			// as this is a syscall whitelist an invalid syscall
-			// is ok and the error can be ignored
-			if (pr_rc == PARSE_INVALID_SYSCALL)
-				continue;
-			die("could not parse line");
-		}
-
-		rc = seccomp_rule_add_exact_array(ctx, SCMP_ACT_ALLOW,
-						  sargs.syscall_nr,
-						  sargs.length, sargs.arg_cmp);
-		if (rc != 0) {
-			rc = seccomp_rule_add_array(ctx, SCMP_ACT_ALLOW,
-						    sargs.syscall_nr,
-						    sargs.length,
-						    sargs.arg_cmp);
-			if (rc != 0) {
-				fprintf(stderr,
-					"seccomp_rule_add_array failed with %i for '%s'\n",
-					rc, buf);
-				errno = 0;
-				die("aborting");
-			}
-		}
-	}
-
- out:
-	if (f != NULL) {
-		if (fclose(f) != 0)
-			die("could not close seccomp file");
-	}
-	sc_map_destroy();
-	return ctx;
-}
-
-void sc_load_seccomp_context(scmp_filter_ctx ctx)
-{
-	int rc;
+	// validate '/' down to profile_path are root-owned and not
+	// 'other' writable to avoid possibility of privilege
+	// escalation via bpf program load when paths are incorrectly
+	// set on the system.
+	validate_bpfpath_is_safe(profile_path);
+
+	// sanity check size/owner
+	struct stat stat_buf;
+	if (stat(profile_path, &stat_buf) < 0)
+		die("cannot stat %s", profile_path);
+	if (stat_buf.st_size > MAX_BPF_SIZE)
+		die("profile %s is too big", profile_path);
```

Comment:
This stat is not strictly needed as the `feof()` check with `fread()` lets us know if it was too big. It's ok to have since it allows us to short-circuit quickly, but I don't think that short circuit is particularly interesting since snapd manages the bpf files and they should be of reasonable size already (again, the latter check ensure security already, so the above stat is just optimization).